assets/js/autofill.js

```diff
@@ -52,169 +52,220 @@ jQuery(document).ready(function( $ ) {
-        jQuery('#shipping_address_2').val(place.address_components[i]['long_name']);
-      } else if(addressType == 'sublocality_level_3'){
-        jQuery('#shipping_address_2').val(place.address_components[i]['long_name']);
-      } else if(addressType == 'sublocality_level_2'){
-        jQuery('#shipping_address_2').val(place.address_components[i]['long_name']);
-      }
-
-      // filling location
-      if(addressType == 'locality'){
-        jQuery('#shipping_city').val(place.address_components[i]['long_name']);
-      }
-      // filling postal code
-      if(addressType == 'postal_code'){
-        jQuery('#shipping_postcode').val(place.address_components[i]['long_name']);
-      }
-
-    }
-
-    if( wcaf.enable_shipping_company_name ) {
-      if(place.hasOwnProperty("name") && place.name) {
-        jQuery('#shipping_company').val(place.name);
-      }
-    }
-
-  }
-
-
-// Filling the address
-  function fillInBillingAddress(){
+// Filling the address. Used both for billing and shipping.
+// @param field string Either 'shipping' or 'billing' fields section.
+	function fillAddress(field){
     place = autofill.getPlace();
-    jQuery('#billing_postcode').val('');
-    jQuery('#billing_address_2').val('');
-    jQuery('#billing_address_1').val('');
-    jQuery('#billing_city').val('');
-    // jQuery('#billing_phone').val('');
-    // jQuery('#billing_company').val('');
 
+		/**
+		* Define all possible data to collect from getPlace() API call.
+		*
+		* Let's use this object to map later on data into each checkout field with
+		* logic tailored to each field.
+		*
+		* See: https://developers.google.com/places/web-service/supported_types
+		*/
+		var address_data = {
+			street_address: '',
+			street_number: '',
+			sublocality: '',
+			route: '',
+			neighborhood: '',
+			floor: '',
+			sublocality_level_1: '',
+			sublocality_level_2: '',
+			sublocality_level_3: '',
+			postal_code: '',
+			postal_code_prefix: '',
+			postal_town: '',
+			locality: '',
+			country: '',
+			administrative_area_level_1: '',
+			administrative_area_level_2: ''
+		};
+
+		/**
+		 * Helper arrays to extract data from API response. The value can came
+		 * in 'long_name' or 'short_name' object properties in the object from
+		 * API's response. This arrays are used to extract the right value form
+		 * the API's response.
+		 */
+		var components_short_name = [
+			'country',
+			'administrative_area_level_1',
+			'administrative_area_level_2'
+		];
+		var components_long_name = [
+			'street_number',
+			'sublocality',
+			'route',
+			'neighborhood',
+			'sublocality_level_1',
+			'sublocality_level_3',
+			'sublocality_level_2',
+			'locality',
+			'postal_code',
+			'postal_code_prefix',
+			'postal_town',
+			'street_address',
+			'floor'
+		];
+
+		// Extract data from API response into address_data object.
     for (var i = 0; i < place.address_components.length; i++) {
       var addressType = place.address_components[i].types[0];
```

Comment:
`place.address_components[i].types` is an array. We have to check if our address type is present in the array. The first element does not always ensure the right parsing. For example, try for `Teston Bridge Country Park, Teston Lane, Teston, Maidstone, UK`